In [1]:
import pandas as pd
import re
import requests
from bs4 import BeautifulSoup as bsp

In [2]:
pattern = re.compile(r'codigoCvm=\d+')
numbers = {}
with open('bovespa.txt', 'r', encoding='utf-8', errors='ignore') as file:
    f = file.read()
    numbers['code'] = re.findall(pattern, f)
    df= pd.DataFrame(numbers)
df = df.drop_duplicates()
df['code'] = df['code'].map(lambda x: x.strip('codigoCvm='))

In [3]:
df.head()

code
0  25291
2  16284
4  21725
6  25283
8  18970

In [4]:
df.to_csv('codigos_cvm.csv', index=False)

In [5]:
codigos = pd.read_csv('codigos_cvm.csv')

In [6]:
codigos.head()

code
0  25291
1  16284
2  21725
3  25283
4  18970

In [7]:
codigos['code'][0]

25291

In [8]:
for i in codigos['code'][:20]:
    print(i)

25291
16284
21725
25283
18970
22179
16705
9954
21032
22357
24953
10456
22217
18066
21490
23264
24961
3050
23248
22349


code = '25291'
url_relatorios = f'http://bvmf.bmfbovespa.com.br/cias-listadas/empresas-listadas/\
ResumoDemonstrativosFinanceiros.aspx?codigoCvm={code}&idioma=pt-br'
url_relatorios

In [67]:
companies = {}
for i in codigos['code'][359:]:
    url_relatorios = f'http://bvmf.bmfbovespa.com.br/cias-listadas/empresas-listadas/ResumoDemonstrativosFinanceiros.aspx?codigoCvm={i}&idioma=pt-br'
    response = requests.get(url_relatorios)
    soup = bsp(response.content, "html.parser")
    content_one = soup.find('div', id='containerDemonstrativos')
    if content_one == None:
        continue
    else:
        content = content_one.find_all('a')
        name = soup.find('span', id='ctl00_contentPlaceHolderConteudo_MenuEmpresasListadas1_lblNomeEmpresa').string
        companies[name] = []
        for i in content[0:3]:
            companies[name].append(i.get('href').lstrip("javascript:AbreFormularioCadastral('").rstrip("')"))

companies

{'SANTHER FAB DE PAPEL STA THEREZINHA S.A.': ['http://www.rad.cvm.gov.br/ENETCONSULTA/frmGerenciaPaginaFRE.aspx?NumeroSequencialDocumento=98603&CodigoTipoInstituicao=2',
  'http://www.rad.cvm.gov.br/ENETCONSULTA/frmGerenciaPaginaFRE.aspx?NumeroSequencialDocumento=96612&CodigoTipoInstituicao=2',
  'http://www.rad.cvm.gov.br/ENETCONSULTA/frmGerenciaPaginaFRE.aspx?NumeroSequencialDocumento=94351&CodigoTipoInstituicao=2'],
 'SANTO ANTONIO ENERGIA S.A.': ['http://www.rad.cvm.gov.br/ENETCONSULTA/frmGerenciaPaginaFRE.aspx?NumeroSequencialDocumento=98343&CodigoTipoInstituicao=2',
  'http://www.rad.cvm.gov.br/ENETCONSULTA/frmGerenciaPaginaFRE.aspx?NumeroSequencialDocumento=95943&CodigoTipoInstituicao=2',
  'http://www.rad.cvm.gov.br/ENETCONSULTA/frmGerenciaPaginaFRE.aspx?NumeroSequencialDocumento=93755&CodigoTipoInstituicao=2'],
 'SANTOS BRASIL PARTICIPACOES S.A.': ['http://www.rad.cvm.gov.br/ENETCONSULTA/frmGerenciaPaginaFRE.aspx?NumeroSequencialDocumento=98200&CodigoTipoInstituicao=2',
  'htt

In [68]:
companies_df = pd.DataFrame(companies).T
companies_df.to_csv('url_demonstrativos.csv', mode='a')

In [74]:
old_companies = pd.read_csv('url_demonstrativos.csv')
old_companies.tail(20)

Unnamed: 0  \
332            TPI - TRIUNFO PARTICIP. E INVEST. S.A.   
333                             TRACK & FIELD CO S.A.   
334     TRANSMISSORA ALIANÇA DE ENERGIA ELÉTRICA S.A.   
335               TRIÂNGULO DO SOL AUTO-ESTRADAS S.A.   
336                                       TRISUL S.A.   
337                   TRONOX PIGMENTOS DO BRASIL S.A.   
338                       ULTRAPAR PARTICIPACOES S.A.   
339                  UNICASA INDÚSTRIA DE MÓVEIS S.A.   
340                                       UNIDAS S.A.   
341                            UNIPAR CARBOCLORO S.A.   
342                                         VALE S.A.   
343                               VALID SOLUÇÕES S.A.   
344                                   VIA VAREJO S.A.   
345            VIVER INCORPORADORA E CONSTRUTORA S.A.   
346                            VULCABRAS/AZALEIA S.A.   
347                                          WEG S.A.   
348                                       WETZEL S.A.   
349                                    WHIRLPOOL S.A.   
350  WLM PART. E COMÉRCIO DE MÁQUINAS E VEÍCULOS S.A.   
351                          YDUQS PARTICIPACOES S.A.   

                                                     0  \
332  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
333  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
334  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
335  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
336  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
337  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
338  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
339  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
340  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
341  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
342  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
343  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
344  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
345  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
346  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
347  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
348  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
349  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
350  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
351  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   

                                                     1  \
332  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
333  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
334  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
335  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
336  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
337  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
338  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
339  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
340  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
341  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
342  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
343  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
344  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
345  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
346  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
347  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
348  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
349  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
350  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   
351  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...   

                                                     2  
332  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...  
333  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...  
334  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...  
335  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...  
336  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...  
337  http://www.rad.cvm.gov.br/ENETCONSULTA/frmGere...  
338  htt

In [ ]:
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry


session = requests.Session()
retry = Retry(connect=3, backoff_factor=0.5)
adapter = HTTPAdapter(max_retries=retry)
session.mount('http://', adapter)
session.mount('https://', adapter)

session.get(old_companies['0'][1])